packages/contracts-bedrock/scripts/L2Genesis.s.sol

```diff
@@ -0,0 +1,388 @@
+
+        L2CrossDomainMessenger(Predeploys.L2_CROSS_DOMAIN_MESSENGER).initialize(
+            L1CrossDomainMessenger(mustGetAddress("L1CrossDomainMessengerProxy"))
+        );
+
+        _checkL2CrossDomainMessenger(impl);
+    }
+
+    /// @notice This predeploy is following the saftey invariant #2,
+    ///         because the constructor args are non-static L1 contract
+    ///         addresses that are being read from the deploy config
+    ///         that are set as immutables.
+    /// @dev Because the constructor args are stored as immutables,
+    ///      we don't have to worry about setting storage slots.
+    function _setSequencerFeeVault() internal {
+        SequencerFeeVault vault = new SequencerFeeVault({
+            _recipient: cfg.sequencerFeeVaultRecipient(),
+            _minWithdrawalAmount: cfg.sequencerFeeVaultMinimumWithdrawalAmount(),
+            _withdrawalNetwork: FeeVault.WithdrawalNetwork.L1
+        });
+
+        address impl = _predeployToCodeNamespace(Predeploys.SEQUENCER_FEE_WALLET);
+        console.log("Setting %s implementation at: %s", "SequencerFeeVault", impl);
+        vm.etch(impl, address(vault).code);
+
+        /// Reset so its not included state dump
+        vm.etch(address(vault), "");
+        vm.resetNonce(address(vault));
+
+        _checkSequencerFeeVault(impl);
+    }
+
+    /// @notice This predeploy is following the saftey invariant #1.
+    ///         We're initializing the implementation with `address(0)` so
+    ///         it's not left uninitialized. After `initialize` is called on the
+    ///         proxy to set the storage slot with the expected value.
+    function _setOptimismMintableERC20Factory() internal {
+        address impl =
+            _setImplementationCode(Predeploys.OPTIMISM_MINTABLE_ERC20_FACTORY, "OptimismMintableERC20Factory");
+
+        OptimismMintableERC20Factory(impl).initialize(address(0));
+
+        OptimismMintableERC20Factory(Predeploys.OPTIMISM_MINTABLE_ERC20_FACTORY).initialize(
+            Predeploys.L2_STANDARD_BRIDGE
+        );
+
+        _checkOptimismMintableERC20Factory(impl);
+    }
+
+    /// @notice This predeploy is following the saftey invariant #1.
+    ///         This contract has no initializer.
+    function _setL1BlockNumber() internal {
+        _setImplementationCode(Predeploys.L1_BLOCK_NUMBER, "L1BlockNumber");
+    }
+
+    /// @notice This predeploy is following the saftey invariant #1.
+    ///         This contract has no initializer.
+    function _setGasPriceOracle() internal {
+        _setImplementationCode(Predeploys.GAS_PRICE_ORACLE, "GasPriceOracle");
+    }
+
+    /// @notice This predeploy is following the saftey invariant #3.
+    function _setGovernanceToken() internal {
+        if (!cfg.enableGovernance()) {
+            console.log("Governance not enabled, skipping setting governanace token");
+            return;
+        }
+
+        GovernanceToken token = new GovernanceToken();
```

Comment:
Previously we were using a "set code" style approach for gov token but now we are deploying it. See https://github.com/ethereum-optimism/optimism/blob/35e559aa7faf1556c01949e270ba36a3707a4d92/op-chain-ops/immutables/immutables.go#L183

This could explain why the old style genesis has "empty fields" in it. I looked thru the deps of the contract and dont see any immutables so this is strange

Reply:
`ERC20Permit` passes `name` and a hardcoded version (`"1"`) to `EIP712`'s constructor, but `ERC20Votes` doesn't have a constructor and doesn't pass set a `name` for `ERC20Permit`, so it's just an empty string

Reply:
`name` is hardcoded to `"Optimism"` by `GovernanceToken` constructor:

```solidity
contract GovernanceToken is ERC20Burnable, ERC20Votes, Ownable {
    /// @notice Constructs the GovernanceToken contract.
    constructor() ERC20("Optimism", "OP") ERC20Permit("Optimism") { }
```

`version` is hardcoded to `"1"` by `ERC20Permit` constructor:

```solidity
abstract contract ERC20Permit is ERC20, IERC20Permit, EIP712 {
	/// ....

    /**
     * @dev Initializes the {EIP712} domain separator using the `name` parameter, and setting `version` to `"1"`.
     *
     * It's a good idea to use the same `name` that is defined as the ERC20 token name.
     */
    constructor(string memory name) EIP712(name, "1") {}
```

@tynes I don't think `_CACHED_THIS` needs to be resolved to a specific address? So resolving all these `immutable`s shouldn't be an issue as long as `--chain-id` is used when running the script (which it already has to for the Preinstalls)?

Reply:
Ah this is totally it - so this caught a bug in the L2 genesis generation for `GovernanceToken` - it means that permit may not work for chains that generated an L2 genesis with gov token set. This does not impact OP Mainnet since the contract was deployed manually, not part of genesis

Reply:
Isn't it broken since this equality check: `address(this) == _CACHED_THIS` will fail in:

```solidity
function _domainSeparatorV4() internal view returns (bytes32) {
        if (address(this) == _CACHED_THIS && block.chainid == _CACHED_CHAIN_ID) {
            return _CACHED_DOMAIN_SEPARATOR;
        } else {
            return _buildDomainSeparator(_TYPE_HASH, _HASHED_NAME, _HASHED_VERSION);
        }
    }
```

since `_CACHED_THIS` is `address(0)`

Reply:
It just will hit the bottom branch and always compute the domain separator and always do so with empty arguments. ie it should be `_HASHED_VERSION = 1` i believe and `_HASHED_NAME = keccak256("OP")`

The idea is to save gas by computing it once at deploy time. The check exists to ensure that it doesn't break in the case that the chainid is modified by the core devs. That is unlikely to happen but just in case.

Reply:
We should write a unit test where we compute the domain separator outside of the function and then assert that it matches what is in the contract. We would expect the test to fail using the existing L2 genesis setup then pass with the new L2 genesis setup

Reply:
> The idea is to save gas by computing it once at deploy time.

Pointing out that this gas savings is still broken even with new genesis setup. This is because we're deploying `GovernanceToken` so `_CACHED_THIS` is set to what ever the resulting address is, but we place the deployed bytecode at the `0xc0d3` namespaced address, so `address(this) == _CACHED_THIS` will always fail

Reply:
This is a good point, thank you for pointing this out. I think its fine that the cache breaks on the `address(this)` check for now, we will likely do a big overhaul on L2 contract management in the future to solve this